### ISSUE TITLE
Stick table entries created with more than one "http-request track-sc0" never expire

### ISSUE THREAD

**Issue by brenc** [haproxy/haproxy#1170]
## Detailed description of the problem

We've always used `tcp-request content track-sc0` but have been testing `http-request track-sc0`. It appears that when a request is tracked by two separate `http-request track-sc0` lines using either the same or two or more stick tables, the `use` field gets incremented and the entry never expires even when `exp=0`.

This does not happen when using `tcp-request content track-sc0`. Using this method, the first request to match is the only one that gets tracked and that entry expires when `exp=0` which I believe is the expected behavior.

Similar to [this issue](https://www.mail-archive.com/haproxy@formilux.org/msg23989.html) and [this issue](https://www.mail-archive.com/haproxy@formilux.org/msg38478.html).

## Expected behavior

The request should only be tracked by the first matching line and should be removed from the stick table when `exp=0`.

## Steps to reproduce the behavior

Using the config below, load up any endpoint in a browser. Even one request will increment the `use` field and that entry will never expire.

Here I made just one request and waited for `exp` to decrement to 0. `use` will continue to increment on subsequent requests:

```
# Output of "show table be-test; show table be-test2; show sess all"
# table: be-test, type: ipv6, size:102400, used:1
0x7f3da8ec26a8: key=::ffff:192.168.0.5 use=1 exp=0 http_req_cnt=2

# table: be-test2, type: ipv6, size:102400, used:0

0x7f3da8f636b0: [10/Mar/2021:03:39:50.789197] id=11 proto=tcpv4 source=172.25.0.
1:56016
  flags=0x8, conn_retries=0, srv_conn=0, pend_pos=0 waiting=0
  frontend=GLOBAL (id=0 mode=tcp), listener=? (id=1) addr=172.25.0.3:8100
  backend=<NONE> (id=-1 mode=-)
  server=<NONE> (id=-1)
  task=0x7f3da9aecfc0 (state=0x00 nice=-64 calls=1 rate=1 exp=1h tmask=0x4 age=0
s)
  si[0]=0x7f3da8f63958 (state=EST flags=0x200008 endp0=CS:0x7f3da9ae7f90 exp=<NE
VER> et=0x000 sub=1)
  si[1]=0x7f3da8f639b0 (state=EST flags=0x204018 endp1=APPCTX:0x7f3da9aecb30 exp
=<NEVER> et=0x000 sub=0)
  co0=0x7f3da9ae9660 ctrl=tcpv4 xprt=RAW mux=PASS data=STRM target=LISTENER:0x7f
3da9b352c0
      flags=0x00003300 fd=32 fd.state=21 updt=1 fd.tmask=0x4
      cs=0x7f3da9ae7f90 csf=0x00008200 ctx=0x7f3da9ae7fa8
  app1=0x7f3da9aecb30 st0=9 st1=0 st2=3 applet=<CLI> tmask=0x4 nice=-64 calls=2
rate=1 cpu=0 lat=0
  req=0x7f3da8f636c0 (f=0xc48202 an=0x0 pipe=0 tofwd=-1 total=55)
      an_exp=<NEVER> rex=1h wex=<NEVER>
      buf=0x7f3da8f636c8 data=0x7f3da8ed99d0 o=0 p=0 i=0 size=16384
  res=0x7f3da8f63720 (f=0x80008002 an=0x0 pipe=0 tofwd=-1 total=168)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x7f3da8f63728 data=0x7f3da8ed50c0 o=168 p=168 i=0 size=16384
```

Here's what it looks like when the same request is tracked with three separate stick tables. The entry shows up in the first table as well but expires when expected.

```
# table: be-test, type: ipv6, size:102400, used:0

# table: be-test2, type: ipv6, size:102400, used:1
0x7f96a86035f8: key=::ffff:192.168.0.5 use=9 exp=0 http_req_cnt=9

# table: be-test3, type: ipv6, size:102400, used:1
0x7f96a86036b8: key=::ffff:192.168.0.5 use=9 exp=0 http_req_cnt=9
```

## Do you have an idea how to solve the issue?

Workaround is to use `tcp-request content track-sc0` or `http-request track-sc1`+ for requests that may match multiple lines (or get more specific with ACLs so requests never match more than one track line).

## What is your configuration?
```
global
    log stdout format raw local0
    maxconn 20000
    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
    ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets
    ssl-default-server-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
    ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets
    stats socket :8100
    stats timeout 1h

defaults
    balance roundrobin
    backlog 10000
    log global
    mode http

    option contstats
    option dontlog-normal
    option dontlognull
    option log-separate-errors
    option httplog

    timeout client 30s
    timeout client-fin 30s
    timeout connect 5s
    timeout http-keep-alive 5s
    timeout http-request 5s
    timeout queue 10s
    timeout server 30s
    timeout tarpit 5s
    timeout tunnel 30s

    errorfile 400 /usr/local/etc/haproxy/errors/400.http
    errorfile 403 /usr/local/etc/haproxy/errors/403.http
    errorfile 408 /usr/local/etc/haproxy/errors/408.http
    errorfile 500 /usr/local/etc/haproxy/errors/500.http
    errorfile 502 /usr/local/etc/haproxy/errors/502.http
    errorfile 503 /usr/local/etc/haproxy/errors/503.http
    errorfile 504 /usr/local/etc/haproxy/errors/504.http

frontend fe-main
    bind :8101 ssl crt /usr/local/etc/haproxy/server.pem alpn h2,http/1.1

    maxconn 10000 
    no option dontlog-normal
    option http-buffer-request
    option forwardfor
    tcp-request inspect-delay 10s

    capture request header Host len 64
    capture request header X-Forwarded-For len 64
    capture request header Referer len 256
  
    # This works as expected.
    #tcp-request content track-sc0 src table be-test
    #tcp-request content track-sc0 src table be-test2

    # Happens when a request is tracked twice within the same stick table.
    http-request track-sc0 src table be-test
    http-request track-sc0 src table be-test

    # Happens with two or more stick tables as well. Any entry beyond the first one to match will never expire.
    # http-request track-sc0 src table be-test
    # http-request track-sc0 src table be-test2

    # This works as expected.
    # http-request track-sc0 src table be-test
    # http-request track-sc1 src table be-test2

    default_backend be-test

backend be-test
    option tcp-smart-connect
    stick-table type ipv6 size 100k expire 5s store http_req_cnt
    server web web:8200 maxconn 128

backend be-test2
    stick-table type ipv6 size 100k expire 5s store http_req_cnt
```

## Output of `haproxy -vv` and `uname -a`

Using 2.2.9 on Debian 10 for production but for this ticket I replicated the issue with the latest `haproxy:2.2-alpine` Docker image:

```
HA-Proxy version 2.2.10-6a09215 2021/03/03 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.10.html
Running on: Linux 4.19.0-14-amd64 #1 SMP Debian 4.19.171-2 (2021-01-30) x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.1j  16 Feb 2021
Running on OpenSSL version : OpenSSL 1.1.1j  16 Feb 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20201203
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
              h2 : mode=HTTP       side=FE|BE     mux=H2
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
	[CACHE] cache
	[FCGI] fcgi-app
```

**Comment by EmericBr** [comment]
Hi brenc,

Does the use counter continue to be inc at each request?

**Comment by EmericBr** [comment]
I suspect the code to re-do a lookup, takes a ref on the table entry  and increases the refcount for each http-request rule, but it may decrease the refcount only at the end of the session and only for the latest entry ref. Doing this, the entry refcount never goes back to 0 and the entry is not removed when expired.

**Comment by EmericBr** [comment]
It would show a constantly increased "use" counter and If you are using the same table, more there is http-request track-sc rules, more the counter will be increased.

**Comment by EmericBr** [comment]
[0001-BUG-MEDIUM-fix-ref-counter-in-table-entry-using-mult.patch.txt](https://github.com/haproxy/haproxy/files/6117511/0001-BUG-MEDIUM-fix-ref-counter-in-table-entry-using-mult.patch.txt)

**Comment by EmericBr** [comment]
This a proposed patch, it applies the same policy than for tcp rules but it seems nor HTTP nor TCP rules are previewed to switch a counter from a table to an other. I need the opinion of @wtarreau

**Comment by brenc** [comment]
@EmericBr yes the use counter increments with each request. Using this config for example:

```
http-request track-sc0 src table be-test
http-request track-sc0 src table be-test
http-request track-sc0 src table be-test
```

The use counter increments twice per request.

**Comment by wtarreau** [comment]
@EmericBr is right, only the first such matching rule must be applied (thus the first one in your case since they're all the same with no condition). This is the documented behavior. However what is not normal is that it's incremented multiple times. I think Emeric's fix is correct, I'll double-check with him.

**Comment by capflam** [comment]
In fact, the fix must only be backported as far as 2.2. The missing test already exists on previous releases.
